docs/04-reference/winglang-plugins.md

```diff
@@ -0,0 +1,76 @@
+---
+title: Compiler Plugins
+id: compiler-plugins
+description: Wing Compiler Plugins Reference
+keywords: [Wing reference, Wing language, language, Wing language spec, Wing programming language, compiler plugins]
+---
+
+Wing compiler plugins are used to extend the Wing compiler. They can be used to add non-functional customizations to the compilation's resulting infrastructure. For example, you can use a compiler plugin to enforce a tagging convention on all resources in the resulting infrastructure.
+
+## Plugin Design
+
+A compiler plugin is a JavaScript file that exports multiple functions. These functions are called by the Wing compiler at various points during compilation also referred to as stages.
+
+At each stage the compiler plugin is provided with a context object that allows the plugin to hook into the compilation process and apply customizations.
+
+A single plugin can implement any number of stages. However, it is recommended to only implement the stages that are needed.
+
+## Compilation Stages
```

Comment:
I would use the term "hooks" instead of "stages". Hooks is more general purpose and doesn't necessarily imply sequential ordering. For example, we want to add a resource factory hook (#1365), and those are not "stages".

"The compiler offers hooks that can be used to customize its behavior at various points during the compilation."